core/src/protosext/mod.rs

```diff
@@ -147,6 +149,11 @@ impl WorkflowActivationExt for WorkflowActivation {
                     variant: Some(workflow_activation_job::Variant::QueryWorkflow(qr))
                 }] if qr.query_id == LEGACY_QUERY_ID)
     }
+    fn attach_build_id_if_needed(&mut self, build_id: &str) {
```

Comment:
This is a bit confusing, why would this logic be on a proto struct, it doesn't know when attaching is needed.
I would put it in the worker logic before it hands over the activation or better yet when the activation is created to avoid mutating unnecessarily.

Reply:
It can't really be when the activation is created without passing the build id down a few layers where it's not currently needed (and would get cloned a bunch of times). I can just inline the method where it's used though.

Reply:
Inlined it w/ comment